### DISCUSSION TITLE
auth.uid() always returns null in local dev DB due to old(?) implementation compared to cloud

### DISCUSSION THREAD

**Discussion by MrHaila** [supabase/supabase#33012] [Troubleshooting]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

We recently noticed that our local development versions of our project started failing tests that previously passed. After looking into it for a few days, we have now narrowed the issue down to the `auth.uid()` call always returning `null` in locally hosted supabase.

## To Reproduce

1. Make a plpgsql function that calls and returns auth.uid()

```
BEGIN
  RETURN auth.uid();
END;
```

2. Invoke it from an authenticated session with something like `const res = await supabase.rpc('debug_info')`

3. Log the results against a cloud deployed environment and a locally run environment. Cloud returns the current session JWT's sub, local returns `null`.

## Expected behavior

auth.uid() should work the same way for local and cloud.

UPDATE:

We nailed down the bug to a differing implementation of auth.uid() in local vs cloud databases.

Self-hosted:
```
CREATE OR REPLACE FUNCTION auth.uid()
 RETURNS uuid
 LANGUAGE sql
 STABLE
AS $function$
  select nullif(current_setting('request.jwt.claim.sub', true), '')::uuid;
$function$
```

Cloud:
```
CREATE OR REPLACE FUNCTION auth.uid()
 RETURNS uuid
 LANGUAGE sql
 STABLE
AS $function$
  select 
  coalesce(
    nullif(current_setting('request.jwt.claim.sub', true), ''),
    (nullif(current_setting('request.jwt.claims', true), '')::jsonb ->> 'sub')
  )::uuid
$function$
```

## Additional context

This did work before for us and seems like there was a regression sometime this summer where DB resets would cause the above, simpler version of the function to be seeded in to fresh local databases.

**Comment by MrHaila** [comment]
For anyone else affected: our current workaround is to add this migration to patch the auth.uid() implementation to be the same locally as it is on the cloud.

```
CREATE OR REPLACE FUNCTION auth.uid()
 RETURNS uuid
 LANGUAGE sql
 STABLE
AS $function$
  select 
  coalesce(
    nullif(current_setting('request.jwt.claim.sub', true), ''),
    (nullif(current_setting('request.jwt.claims', true), '')::jsonb ->> 'sub')
  )::uuid
$function$
```

**Comment by Hallidayo** [comment]
Hi everyone, due to inactivity on this issue I've moved the issue over to discussions/enhancements.
